Shutdown server socket first

TimWSpence opened the pull request:
Opening this for comment. The idea is to shut down the server socket first so we stop accepting new connections and then enter some kind of graceful shutdown period. See https://github.com/typelevel/fs2/issues/2289 for context.

As it stands, this should close the server socket first as desired but will then just interrupt any active connections. We probably want some more refined behaviour here.

@armanbilge and I were chatting and thought the behaviour (if possible) should be something like:
1. stop accepting new connections (close server socket)
2. stop accepting new requests on existing connections (respond with `Connection: close`) and possibly just close open sockets which aren't actively processing a request
3. stop responding, even to in-flight requests (some kind of hard shutdown - I guess timed?)

How does that sound? @ChristopherDavenport in particular might have thoughts?

TimWSpence committed:
Sketch of shutting down Ember TCP server socket before waiting for
active connections to finish

TimWSpence committed:
Fix existing tests

TimWSpence committed:
Remove unnecessary `interruptWhen` as fs2 will now terminate stream when
server socket is closed

TimWSpence committed:
Merge remote-tracking branch 'upstream/series/0.23' into shutdown-server-socket-first

TimWSpence committed:
Scalafix

armanbilge commented on ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala:
```diff
@@ -75,31 +76,53 @@ private[server] object ServerHelpers extends ServerHelpersPlatform {
+            .evalTap(e => ready.complete(e.map(_._1)))
+            .rethrow
+            .allocated
+            .uncancelable
```
Why do we need uncancelable here?

TimWSpence replied:
I cannot remember 😅 Looks like we don't

TimWSpence committed:
Remove unnecessary uncancelable

TimWSpence committed:
Gracefully serving active connections

TimWSpence committed:
Set Connection: close when processing a request after shutdown is initiated

TimWSpence committed:
Hard shutdown after grace period

armanbilge commented on ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala:
```diff
@@ -75,31 +76,52 @@ private[server] object ServerHelpers extends ServerHelpersPlatform {
-      webSocketKey,
-      enableHttp2,
-    )
+    val res = Resource.uncancelable[F, Unit] { _ =>
+      Resource
+        .eval(
```
Can this `eval` block be combined with the one below? And why do we need `Resource.uncancelable`?

And then can `Resource.eval(...).flatMap(... onFinalize(...)` just become `Resource.make(...)(...)` ?

TimWSpence committed:
Coalesce `Resource.eval` blocks

TimWSpence committed:
Simplfy Resource building

TimWSpence committed:
Scalafix :sob:

TimWSpence committed:
Check isShutdown again

Means that we send `Connection: close` for requests that had begun
processing before shutdown signal was set

TimWSpence committed:
Resolve races

TimWSpence committed:
Use stream interruption instead of timeout on request handling

The timeout doesn't appear to be reliable

TimWSpence committed:
Remove hack

TimWSpence committed:
Faster feedback when test fails

TimWSpence committed:
Ensure we wait even if finalizer throws

TimWSpence commented:
@armanbilge apologies but I'm probably not going to have time to finish this until the new year but I'll get back on it then. Or feel free to debug native in my absense 😛

armanbilge commented:
I see what you did there 😂 have a wonderful holiday!

danicheg commented on ember-core/shared/src/main/scala/org/http4s/ember/core/Util.scala:
```diff
@@ -36,7 +36,7 @@ private[ember] object Util extends UtilPlatform {
   private[this] val keepAliveCi = ci"keep-alive"
   private[this] val connectionCi = ci"connection"
-  private[this] val close = Connection(NonEmptyList.of(closeCi))
+  val close: Connection = Connection(NonEmptyList.of(closeCi))
```
Maybe some private context isn't that bad?
```suggestion
  private[http4s] val close: Connection = Connection(NonEmptyList.of(closeCi))
```

danicheg commented on ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala:
```diff
@@ -75,31 +76,40 @@ private[server] object ServerHelpers extends ServerHelpersPlatform {
-      webSocketKey,
-      enableHttp2,
-    )
+    val res = Resource.uncancelable[F, Unit] { _ =>
```
Upvote this question 👍🏻

TimWSpence replied:
So sorry that I'm only getting round to this now. I think I put it in because of the `allocated` but actually resource acquisition (`Resource.make`) is uncancelable anyway, right?

armanbilge replied:
Yes, `make` is uncancelable.

danicheg commented on ember-server/shared/src/main/scala/org/http4s/ember/server/internal/Shutdown.scala:
```diff
@@ -21,12 +21,13 @@ import cats.effect.implicits._
   def await: F[Unit]
   def signal: F[Unit]
+  def isShutdown: F[Boolean]
+  def gracePeriod: Duration
```
What about `closerGracePeriod` or `shutdownGracePeriod`?

danicheg reviewed (commented):
Pretty interesting work! A few comments after a quite surfaced peek

TimWSpence committed:
Merge remote-tracking branch 'upstream/series/0.23' into shutdown-server-socket-first

TimWSpence committed:
Remove unnecessaru uncancelable

TimWSpence committed:
Renaming and limiting public API

TimWSpence committed:
Fix broken merge
